Mathlib/CategoryTheory/Limits/FormalCoproducts.lean

```diff
@@ -0,0 +1,439 @@
+
+## TODO
+
+* `FormalCoproduct.of C : C ⥤ FormalCoproduct.{w} C` probably preserves every limit?
+
+-/
+
+
+universe w w₁ w₂ w₃ v v₁ v₂ v₃ u u₁ u₂ u₃
+
+open Opposite CategoryTheory Functor
+
+namespace CategoryTheory
+
+namespace Limits
+
+variable {C : Type u} [Category.{v} C] (A : Type u₁) [Category.{v₁} A]
+
+variable (C) in
+/-- A formal coproduct is an indexed set of objects, where ⟨I, f⟩ corresponds to the "formal
+coproduct" `⨿ (i : I), f i`, where `f i : C` is the `i`ᵗʰ component. -/
+structure FormalCoproduct where
+  /-- The indexing type. -/
+  I : Type w
+  /-- The object in the original category indexed by `x : I`. -/
+  obj (i : I) : C
+
+namespace FormalCoproduct
+
+/-- A morphism `(⨿ (i : X.I), X.obj i) ⟶ (⨿ (j : Y.I), Y.obj i)` is given by first a function
+on the indexing sets `f : X.I → Y.I`, and then for each `i : X.I` a morphism
+`X.obj i ⟶ Y.obj (f i)`. -/
+structure Hom (X Y : FormalCoproduct.{w} C) where
+  /-- The function on the indexing sets. -/
+  f : X.I → Y.I
+  /-- The map on each component. -/
+  φ (i : X.I) : X.obj i ⟶ Y.obj (f i)
+
+-- this category identifies to the fullsubcategory of the category of
+-- presheaves of sets on `C` which are coproducts of representable presheaves
+@[simps!] instance category : Category (FormalCoproduct.{w} C) where
+  Hom := Hom
+  -- JH: 𝟙 below should be eqToHom so that Lean does not check if the objects are defeq?
+  id X := { f := id, φ := fun _ ↦ 𝟙 _ }
+  comp α β := { f := β.f ∘ α.f, φ := fun _ ↦ α.φ _ ≫ β.φ _ }
+
+@[ext (iff := false)]
+lemma hom_ext {X Y : FormalCoproduct.{w} C} {f g : X ⟶ Y} (h₁ : f.f = g.f)
+    (h₂ : ∀ (i : X.I), f.φ i ≫ eqToHom (by rw [h₁]) = g.φ i) : f = g := by
+  obtain ⟨f, F⟩ := f
+  obtain ⟨g, G⟩ := g
+  obtain rfl : f = g := h₁
+  obtain rfl : F = G := by ext i; simpa using h₂ i
+  rfl
+
+lemma hom_ext_iff {X Y : FormalCoproduct.{w} C} (f g : X ⟶ Y) :
+    f = g ↔ ∃ h₁ : f.f = g.f, ∀ (i : X.I), f.φ i ≫ eqToHom (by rw [h₁]) = g.φ i :=
+  ⟨(· ▸ by simp), fun ⟨h₁, h₂⟩ ↦ hom_ext h₁ h₂⟩
+
+lemma hom_ext_iff' {X Y : FormalCoproduct.{w} C} (f g : X ⟶ Y) :
+    f = g ↔ ∀ i : X.I, ∃ h₁ : f.f i = g.f i, f.φ i ≫ eqToHom (by rw [h₁]) = g.φ i :=
+  ⟨(· ▸ by simp), fun h ↦ hom_ext (funext fun i ↦ (h i).fst) fun i ↦ (h i).snd⟩
+
+/-- A way to create isomorphisms in the category of formal coproducts, by creating an `Equiv`
+between the indexing sets, and then correspondingly isomorphisms of each component. -/
+@[simps!] def isoOfComponents {X Y : FormalCoproduct.{w} C} (e : X.I ≃ Y.I)
+    (h : ∀ i, X.obj i ≅ Y.obj (e i)) : X ≅ Y where
+  hom := { f := e, φ := fun i ↦ (h i).hom }
+  inv := { f := e.symm, φ := fun i ↦ eqToHom (by simp) ≫ (h (e.symm i)).inv }
+  hom_inv_id := by ext <;> aesop
+  inv_hom_id := by ext <;> aesop
+
+variable (C) in
+/-- An object of the original category produces a formal coproduct on that object only, so indexed
+by `PUnit`, the type with one element. -/
+@[simps!] def of : C ⥤ FormalCoproduct.{w} C where
```

Comment:
```suggestion
@[simps!] def incl : C ⥤ FormalCoproduct.{w} C where
```

See below, but I think that some of the names get unnecessarily confusing because of the name of this functor (since we use `of` in some names to state some of the hypothesis). I think `incl` is both short enough and way less confusing.

Reply:
done